In [12]:
import pandas as pd

# Data Wrangling
The data for this project comes from Kaggle, with all data coming from 3 csv files.

1. nfl_teams
    - Contains team data past and present
2. nfl_stadiums
    - Contains stadium data past and present
3. spreadspoke_scores
    - Contains score and weather data for all NFL games

In [ ]:
## Importing csv files into panda DataFrames

In [77]:
filenameTeam = '/Users/Rezalution/Desktop/nfl-scores-and-betting-data/nfl_teams.csv'
filenameStadiums = '/Users/Rezalution/Desktop/nfl-scores-and-betting-data/nfl_stadiums.csv'
filenameGames = '/Users/Rezalution/Desktop/nfl-scores-and-betting-data/spreadspoke_scores.csv'

In [78]:
nflTeams = pd.read_csv(filenameTeam,encoding ='latin1')
nflStadiums = pd.read_csv(filenameStadiums,encoding ='latin1')
nflGames = pd.read_csv(filenameScores,encoding ='latin1')

In [ ]:
## Top view of the 3 DataFrames

In [7]:
nflTeams.head()

team_name team_name_short team_id team_id_pfr team_conference  \
0    Arizona Cardinals       Cardinals     ARI         CRD             NFC   
1    Phoenix Cardinals       Cardinals     ARI         CRD             NFC   
2  St. Louis Cardinals       Cardinals     ARI         ARI             NFC   
3      Atlanta Falcons         Falcons     ATL         ATL             NFC   
4     Baltimore Ravens          Ravens     BAL         RAV             AFC   

  team_division team_conference_pre2002 team_division_pre2002  
0      NFC West                     NFC              NFC West  
1           NaN                     NFC              NFC East  
2           NaN                     NFC              NFC East  
3     NFC South                     NFC              NFC West  
4     AFC North                     AFC           AFC Central

In [8]:
nflStadiums.head()

stadium_name   stadium_location  stadium_open  stadium_close  \
0         Alamo Dome    San Antonio, TX           NaN            NaN   
1     Alltel Stadium   Jacksonville, FL           NaN            NaN   
2     Alumni Stadium  Chestnut Hill, MA           NaN            NaN   
3    Anaheim Stadium        Anaheim, CA        1980.0         1994.0   
4  Arrowhead Stadium    Kansas City, MO        1972.0            NaN   

  stadium_type                           stadium_address  \
0       indoor     100 Montana St, San Antonio, TX 78203   
1          NaN                                       NaN   
2      outdoor     Perimeter Rd, Chestnut Hill, MA 02467   
3      outdoor  2000 E Gene Autry Way, Anaheim, CA 92806   
4      outdoor     1 Arrowhead Dr, Kansas City, MO 64129   

  stadium_weather_station_code stadium_weather_type stadium_capacity  \
0                        78203                 dome            72000   
1                          NaN                  NaN              NaN   
2                        02467                 cold              NaN   
3                        92806                 warm              NaN   
4                        64129                 cold           76,416   

  stadium_surface      STATION                       NAME  LATITUDE  \
0       FieldTurf          NaN                        NaN       NaN   
1             NaN          NaN                        NaN       NaN   
2           Grass          NaN                        NaN       NaN   
3             NaN          NaN                        NaN       NaN   
4           Grass  US1MOJC0028  KANSAS CITY 5.1 SE, MO US   39.0692   

   LONGITUDE  ELEVATION  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4   -94.4871      264.9

In [80]:
nflGames.head()

schedule_date  schedule_season schedule_week  schedule_playoff  \
0    09/02/1966             1966             1             False   
1    09/03/1966             1966             1             False   
2    09/04/1966             1966             1             False   
3    09/09/1966             1966             2             False   
4    09/10/1966             1966             1             False   

            team_home  score_home  score_away        team_away  \
0      Miami Dolphins        14.0        23.0  Oakland Raiders   
1      Houston Oilers        45.0         7.0   Denver Broncos   
2  San Diego Chargers        27.0         7.0    Buffalo Bills   
3      Miami Dolphins        14.0        19.0    New York Jets   
4   Green Bay Packers        24.0         3.0  Baltimore Colts   

  team_favorite_id  spread_favorite over_under_line         stadium  \
0              NaN              NaN             NaN     Orange Bowl   
1              NaN              NaN             NaN    Rice Stadium   
2              NaN              NaN             NaN  Balboa Stadium   
3              NaN              NaN             NaN     Orange Bowl   
4              NaN              NaN             NaN   Lambeau Field   

   stadium_neutral  weather_temperature  weather_wind_mph weather_humidity  \
0            False                 83.0               6.0               71   
1            False                 81.0               7.0               70   
2            False                 70.0               7.0               82   
3            False                 82.0              11.0               78   
4            False                 64.0               8.0               62   

  weather_detail  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN

# Data Preparation

## Data Loading

In [74]:
## DataFrame info

### NFL Teams

In [75]:
nflTeams.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41 entries, 0 to 40
Data columns (total 8 columns):
team_name                  41 non-null object
team_name_short            41 non-null object
team_id                    41 non-null object
team_id_pfr                41 non-null object
team_conference            41 non-null object
team_division              33 non-null object
team_conference_pre2002    41 non-null object
team_division_pre2002      39 non-null object
dtypes: object(8)
memory usage: 2.7+ KB


The NFL Team dataframe contains 8 columns with 41 rows. There are 2 columns that contain NULL values (team_division and team_division_pre2002). The NULL values will not need to be editted nor the rows dropped because divisional data will not be used in the data analysis.  

### NFL Stadiums

In [27]:
nflStadiums.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 15 columns):
stadium_name                    100 non-null object
stadium_location                100 non-null object
stadium_open                    76 non-null float64
stadium_close                   41 non-null float64
stadium_type                    93 non-null object
stadium_address                 91 non-null object
stadium_weather_station_code    90 non-null object
stadium_weather_type            93 non-null object
stadium_capacity                39 non-null object
stadium_surface                 53 non-null object
STATION                         52 non-null object
NAME                            52 non-null object
LATITUDE                        52 non-null float64
LONGITUDE                       52 non-null float64
ELEVATION                       52 non-null float64
dtypes: float64(5), object(10)
memory usage: 11.8+ KB


The NFL Stadium dataframe contains 15 columns with 100 rows. There are 13 columns that contain NULL values, with only 2 columns that contain full data. Futher investgating will need to be done to determine the correct actions for these NULL values.

### NFL Game Details

In [21]:
nflScore.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12667 entries, 0 to 12666
Data columns (total 17 columns):
schedule_date          12667 non-null object
schedule_season        12667 non-null int64
schedule_week          12667 non-null object
schedule_playoff       12667 non-null bool
team_home              12667 non-null object
score_home             12411 non-null float64
score_away             12411 non-null float64
team_away              12667 non-null object
team_favorite_id       9932 non-null object
spread_favorite        9932 non-null float64
over_under_line        9922 non-null object
stadium                12667 non-null object
stadium_neutral        12667 non-null bool
weather_temperature    11934 non-null float64
weather_wind_mph       11934 non-null float64
weather_humidity       8388 non-null object
weather_detail         2637 non-null object
dtypes: bool(2), float64(5), int64(1), object(9)
memory usage: 1.5+ MB


The NFL Game Details dataframe contains 17 columns with 12667 rows. There are 9 columns that contain NULL values. Just like the NFL Stadium Datarame above futher investgating will need to be done.

## Data Cleaning

### Stadiums

In [85]:
nflStadiums.head()

stadium_name   stadium_location  stadium_open  stadium_close  \
0         Alamo Dome    San Antonio, TX           NaN            NaN   
1     Alltel Stadium   Jacksonville, FL           NaN            NaN   
2     Alumni Stadium  Chestnut Hill, MA           NaN            NaN   
3    Anaheim Stadium        Anaheim, CA        1980.0         1994.0   
4  Arrowhead Stadium    Kansas City, MO        1972.0            NaN   

  stadium_type                           stadium_address  \
0       indoor     100 Montana St, San Antonio, TX 78203   
1          NaN                                       NaN   
2      outdoor     Perimeter Rd, Chestnut Hill, MA 02467   
3      outdoor  2000 E Gene Autry Way, Anaheim, CA 92806   
4      outdoor     1 Arrowhead Dr, Kansas City, MO 64129   

  stadium_weather_station_code stadium_weather_type stadium_capacity  \
0                        78203                 dome            72000   
1                          NaN                  NaN              NaN   
2                        02467                 cold              NaN   
3                        92806                 warm              NaN   
4                        64129                 cold           76,416   

  stadium_surface      STATION                       NAME  LATITUDE  \
0       FieldTurf          NaN                        NaN       NaN   
1             NaN          NaN                        NaN       NaN   
2           Grass          NaN                        NaN       NaN   
3             NaN          NaN                        NaN       NaN   
4           Grass  US1MOJC0028  KANSAS CITY 5.1 SE, MO US   39.0692   

   LONGITUDE  ELEVATION  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4   -94.4871      264.9

In [71]:
nflStadiumsCleaned = nflStadiums.drop(columns=['stadium_open','stadium_close','stadium_address','stadium_weather_station_code','stadium_capacity','STATION','NAME','LATITUDE','LONGITUDE'])

In [72]:
nflStadiumsCleaned.head()

stadium_name   stadium_location stadium_type stadium_weather_type  \
0         Alamo Dome    San Antonio, TX       indoor                 dome   
1     Alltel Stadium   Jacksonville, FL          NaN                  NaN   
2     Alumni Stadium  Chestnut Hill, MA      outdoor                 cold   
3    Anaheim Stadium        Anaheim, CA      outdoor                 warm   
4  Arrowhead Stadium    Kansas City, MO      outdoor                 cold   

  stadium_surface  ELEVATION  
0       FieldTurf        NaN  
1             NaN        NaN  
2           Grass        NaN  
3             NaN        NaN  
4           Grass      264.9

In [83]:
nflStadiumsCleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 6 columns):
stadium_name            100 non-null object
stadium_location        100 non-null object
stadium_type            93 non-null object
stadium_weather_type    93 non-null object
stadium_surface         53 non-null object
ELEVATION               52 non-null float64
dtypes: float64(1), object(5)
memory usage: 4.8+ KB


Columns that are not neccessary for the analysis were dropped off the DataFrame. This brought the column size to 6 from 15. The values that are NULL in the remaining columns will remain the same. Changing the NULL values to integer 0 will not work because most of the values are of type object and it would not make sense to have it as a integer value. The only numercial column is Elevation, but changing the NULL values to 0 will mean that those entries are at sea-level instead of not having a value. 

In [84]:
nflGames.head()

schedule_date  schedule_season schedule_week  schedule_playoff  \
0    09/02/1966             1966             1             False   
1    09/03/1966             1966             1             False   
2    09/04/1966             1966             1             False   
3    09/09/1966             1966             2             False   
4    09/10/1966             1966             1             False   

            team_home  score_home  score_away        team_away  \
0      Miami Dolphins        14.0        23.0  Oakland Raiders   
1      Houston Oilers        45.0         7.0   Denver Broncos   
2  San Diego Chargers        27.0         7.0    Buffalo Bills   
3      Miami Dolphins        14.0        19.0    New York Jets   
4   Green Bay Packers        24.0         3.0  Baltimore Colts   

  team_favorite_id  spread_favorite over_under_line         stadium  \
0              NaN              NaN             NaN     Orange Bowl   
1              NaN              NaN             NaN    Rice Stadium   
2              NaN              NaN             NaN  Balboa Stadium   
3              NaN              NaN             NaN     Orange Bowl   
4              NaN              NaN             NaN   Lambeau Field   

   stadium_neutral  weather_temperature  weather_wind_mph weather_humidity  \
0            False                 83.0               6.0               71   
1            False                 81.0               7.0               70   
2            False                 70.0               7.0               82   
3            False                 82.0              11.0               78   
4            False                 64.0               8.0               62   

  weather_detail  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN

### Game Details

In [86]:
nflGames.head()

schedule_date  schedule_season schedule_week  schedule_playoff  \
0    09/02/1966             1966             1             False   
1    09/03/1966             1966             1             False   
2    09/04/1966             1966             1             False   
3    09/09/1966             1966             2             False   
4    09/10/1966             1966             1             False   

            team_home  score_home  score_away        team_away  \
0      Miami Dolphins        14.0        23.0  Oakland Raiders   
1      Houston Oilers        45.0         7.0   Denver Broncos   
2  San Diego Chargers        27.0         7.0    Buffalo Bills   
3      Miami Dolphins        14.0        19.0    New York Jets   
4   Green Bay Packers        24.0         3.0  Baltimore Colts   

  team_favorite_id  spread_favorite over_under_line         stadium  \
0              NaN              NaN             NaN     Orange Bowl   
1              NaN              NaN             NaN    Rice Stadium   
2              NaN              NaN             NaN  Balboa Stadium   
3              NaN              NaN             NaN     Orange Bowl   
4              NaN              NaN             NaN   Lambeau Field   

   stadium_neutral  weather_temperature  weather_wind_mph weather_humidity  \
0            False                 83.0               6.0               71   
1            False                 81.0               7.0               70   
2            False                 70.0               7.0               82   
3            False                 82.0              11.0               78   
4            False                 64.0               8.0               62   

  weather_detail  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN

In [96]:
nflGamesCleaned = nflGames.drop(columns=['over_under_line','spread_favorite','weather_detail'])

In [97]:
nflGamesCleaned.head()

schedule_date  schedule_season schedule_week  schedule_playoff  \
0    09/02/1966             1966             1             False   
1    09/03/1966             1966             1             False   
2    09/04/1966             1966             1             False   
3    09/09/1966             1966             2             False   
4    09/10/1966             1966             1             False   

            team_home  score_home  score_away        team_away  \
0      Miami Dolphins        14.0        23.0  Oakland Raiders   
1      Houston Oilers        45.0         7.0   Denver Broncos   
2  San Diego Chargers        27.0         7.0    Buffalo Bills   
3      Miami Dolphins        14.0        19.0    New York Jets   
4   Green Bay Packers        24.0         3.0  Baltimore Colts   

  team_favorite_id         stadium  stadium_neutral  weather_temperature  \
0              NaN     Orange Bowl            False                 83.0   
1              NaN    Rice Stadium            False                 81.0   
2              NaN  Balboa Stadium            False                 70.0   
3              NaN     Orange Bowl            False                 82.0   
4              NaN   Lambeau Field            False                 64.0   

   weather_wind_mph weather_humidity  
0               6.0               71  
1               7.0               70  
2               7.0               82  
3              11.0               78  
4               8.0               62

In [98]:
nflGamesCleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12667 entries, 0 to 12666
Data columns (total 14 columns):
schedule_date          12667 non-null object
schedule_season        12667 non-null int64
schedule_week          12667 non-null object
schedule_playoff       12667 non-null bool
team_home              12667 non-null object
score_home             12411 non-null float64
score_away             12411 non-null float64
team_away              12667 non-null object
team_favorite_id       9932 non-null object
stadium                12667 non-null object
stadium_neutral        12667 non-null bool
weather_temperature    11934 non-null float64
weather_wind_mph       11934 non-null float64
weather_humidity       8388 non-null object
dtypes: bool(2), float64(4), int64(1), object(7)
memory usage: 1.2+ MB


Columns that are not neccessary for the analysis were dropped off the DataFrame. This brought the column size to 14 from 17. The values that are NULL in the remaining columns will remain the same. Changing the NULL values to integer 0 will not work because most of the values are of type object and it would not make sense to have it as a integer value. Changing the NULL values for the numerical values to 0 will skew any of the statistical anaylsis done on those entries since those values will be calculated as 0. 

In [115]:
nflGamesCleaned[pd.isnull(nflGamesCleaned['weather_temperature'])].groupby(by='schedule_season').count()

schedule_date  schedule_week  schedule_playoff  team_home  \
schedule_season                                                              
1966                        41             41                41         41   
1967                        58             58                58         58   
1968                        65             65                65         65   
1969                        66             66                66         66   
1979                         8              8                 8          8   
1980                         8              8                 8          8   
1981                         8              8                 8          8   
1982                        14             14                14         14   
1983                         8              8                 8          8   
1984                         8              8                 8          8   
1985                         8              8                 8          8   
1986                         9              9                 9          9   
1987                         7              7                 7          7   
1988                         8              8                 8          8   
1989                         7              7                 7          7   
1990                        11             11                11         11   
1991                         7              7                 7          7   
1992                         9              9                 9          9   
1993                         8              8                 8          8   
1994                        10             10                10         10   
1995                        11             11                11         11   
1996                        10             10                10         10   
1997                        11             11                11         11   
1998                         8              8                 8          8   
1999                         5              5                 5          5   
2000                         9              9                 9          9   
2001                         8              8                 8          8   
2002                        11             11                11         11   
2003                         8              8                 8          8   
2004                         9              9                 9          9   
2005                         9              9                 9          9   
2006                         8              8                 8          8   
2007                         9              9                 9          9   
2008                         8              8                 8          8   
2009                         5              5                 5          5   
2010                         9              9                 9          9   
2011                         9              9                 9          9   
2012                         8              8                 8          8   
2013                        10             10                10         10   
2019                       200            200               200        200   

                 score_home  score_away  team_away  team_favorite_id  stadium  \
schedule_season                                                                 
1966                     41          41         41                 0       41   
1967                     58          58         58                 0       58   
1968                     65          65         65                 0       65   
1969                     66          66         66                 0       66   
1979                      8           8          8                 8        8   
1980                      8           8          8                 8        8   
1981                      8           8          8                 8        8   
1982       

In [119]:
nflGamesCleaned = nflGamesCleaned.dropna(axis=0,subset=['weather_temperature'])

In [120]:
nflGamesCleaned.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11934 entries, 0 to 12658
Data columns (total 14 columns):
schedule_date          11934 non-null object
schedule_season        11934 non-null int64
schedule_week          11934 non-null object
schedule_playoff       11934 non-null bool
team_home              11934 non-null object
score_home             11878 non-null float64
score_away             11878 non-null float64
team_away              11934 non-null object
team_favorite_id       9629 non-null object
stadium                11934 non-null object
stadium_neutral        11934 non-null bool
weather_temperature    11934 non-null float64
weather_wind_mph       11934 non-null float64
weather_humidity       8388 non-null object
dtypes: bool(2), float64(4), int64(1), object(7)
memory usage: 1.2+ MB


A majority of the missing weather data comes from 1966-1969 seasons as well as 200 missing weather entries from 2019. Because weather is a critical factor in this anaylsis, the rows with NULL values will be dropped. This should not effect the outcome since the percentage of NULL entries only around 5.8%.

## Outliers

In [123]:
nflStadiumsCleaned.describe()

ELEVATION
count    52.000000
mean    181.519231
std     310.972267
min       1.800000
25%      12.250000
50%     154.400000
75%     212.525000
max    1611.200000

In [124]:
nflGamesCleaned.describe()

schedule_season    score_home    score_away  weather_temperature  \
count     11934.000000  11878.000000  11878.000000         11934.000000   
mean       1994.479219     22.282287     19.536286            58.618988   
std          14.882158     10.536203     10.125390            15.539685   
min        1966.000000      0.000000      0.000000            -6.000000   
25%        1982.000000     14.000000     13.000000            48.000000   
50%        1996.000000     21.000000     19.500000            61.000000   
75%        2007.000000     29.000000     27.000000            72.000000   
max        2019.000000     72.000000     62.000000            97.000000   

       weather_wind_mph  
count      11934.000000  
mean           7.853276  
std            5.627164  
min            0.000000  
25%            4.000000  
50%            8.000000  
75%           12.000000  
max           40.000000

There were not outliers found within the data. The biggest concern would be if weather temperature had an extreme min or max value but the min and max fit within the norms. The NFL seasons also fall within the range of 1966 - 2019. Also there are not outliersin the game score field, with the values looking normal for a NFL game. 